include/deal.II/optimization/line_minimization.h

```diff
@@ -0,0 +1,592 @@
+   * within the provided @p bounds use quadratic fit. Fallback to bisection
+   * if quadratic fit fails.
+   */
+  template <typename NumberType>
+  NumberType
+  poly_fit(const NumberType                        x_low,
+           const NumberType                        f_low,
+           const NumberType                        g_low,
+           const NumberType                        x_hi,
+           const NumberType                        f_hi,
+           const NumberType                        g_hi,
+           const NumberType                        x_rec,
+           const NumberType                        f_rec,
+           const unsigned int                      i,
+           const std::pair<NumberType, NumberType> bounds);
+
+  /**
+   * Same as above but does cubic fit using three points.
+   */
+  template <typename NumberType>
+  NumberType
+  poly_fit_three_points(const NumberType                        x_low,
+                        const NumberType                        f_low,
+                        const NumberType                        g_low,
+                        const NumberType                        x_hi,
+                        const NumberType                        f_hi,
+                        const NumberType                        g_hi,
+                        const NumberType                        x_rec,
+                        const NumberType                        f_rec,
+                        const unsigned int                      i,
+                        const std::pair<NumberType, NumberType> bounds);
+
+
+  /**
+   * Perform a line search in $(0,max]$ with strong Wolfe conditions
+   * \f[
+   * f(\alpha) \le f(0) + \alpha \mu f'(0) \\
+   * |f'(\alpha)| \le \eta |f'(0)|
+   * \f]
+   * using one dimensional
+   * functions @p func and a function @p interpolate to choose a new point
+   * from the interval based on the function values and derivatives at its ends.
+   * @p a1 is a trial estimate of the step.
+   *
+   * The function implements Algorithms
+   * 2.6.2 and 2.6.4 on pages 34-35 in Fletcher, 2013, Practical methods of
+   * optimization. these are minor variations of the Algorithm 3.5 and 3.6 on
+   * pages 60-61 in Nocedal and Wright, Numerical optimization.
+   * It consists of a bracketing phase and a zoom phase, where @p interpolate is used.
+   *
+   * The function returns the step size and the number of times function @p func was called.
+   *
+   * @param func a one dimensional function which returns value and derivative
+   * at the given point.
+   * @param f0 function value the origin
+   * @param g0 function derivative the origin
+   * @param interpolate a function which determines how interpolation is done
+   * during the zoom phase. It takes values and derivatives at the current
+   * interval/bracket ($f\_low < f\_hi$) as well as the function value at the
+   * previous step. The iteration count and the bounds on the interval are also
+   * provided.
```

Comment:
It'd probably be good to mention which of the `<X>_fit` functions above can be used as is for interpolation.

Reply:
done (above).